### DISCUSSION TITLE
Input for powershell script

### DISCUSSION THREAD

**Discussion by aniketgm** [VladimirMarkelov/ttdl#84] [General]
I'm trying custom formatting. Let's say, I have the following task:
```
# P Spent  Project Subject
---------------------------
2          myapp   Prepare design for Scheduler feature @ui due:2023-06-24
---------------------------
1 todos (of 2 total)
```
And I have written a simple powershell script for removing due date:

![image](https://github.com/VladimirMarkelov/ttdl/assets/14215698/5afbc384-2bbe-4c0a-bc82-60629f965ab3)

But when I add the tag -- `!hidedue` I'm get the following error. Powershell uses a variable `$Args` for command line inputs. Not sure how input is passed here. What is wrong or am I missing anything ? Can you please share some insight ?

![image](https://github.com/VladimirMarkelov/ttdl/assets/14215698/d240e0e4-7648-47f8-8494-078e4cb5f7eb)

**Comment by VladimirMarkelov** [comment]
Originally, the feature was done for a person who worked on Linux. When it got good enough for him, we marked the feature done, I did cursory test on Windows(mostly using cmd instead of powershell) and that was all. 

It looks like you have figured out some quirks of the feature, but just in case I'll name everything I did to made it work somehow:

1. First, uncomment line `script_ext = ".ps1"` in the configuration file, otherwise TTDL won't find any file ending with `.ps1`
2. As the number of tags can be big, I do not pass the JSON as an argument of a script. TTDL runs the script with opened pipe to STDIN, so the script has to read from `$input` in Powershell. I'll add a small example at the end of my comment.
3. Due to bullet point 2, the original advise to run Windows scripts with "powershell -c" does not work at all: `$input` is always empty. So, you have to replace `shell = ["powershell", "-c"]` with `shell = ["powershell", "-F"]` in the configuration file
4. I encountered a few bugs: a) setting a tag value to empty string should make TTDL remove the tag from the description before printing, but it does not; b) moreover, it adds duplicated tag with a new value b) changing some tags like `due` changes the text printed in the corresponding column, but in the description the tag value remains unchanged

Example:

Let's assume `todo.txt` contains the line:

```
test2 due:2023-06-15 spent:40 !hidedue:abc
```

And the script `ttdl-hidedue.ps1`:

```powershell
$obj = $input | ConvertFrom-Json
$obj.specialTags | Where {$_.due} | ForEach{$_.due = ''}
$obj.specialTags | Where {$_."!hidedue"} | ForEach{$_."!hidedue" = '????'}
Write-Output (ConvertTo-Json $obj)
```

Now list tasks:

```
ttdl list
# D P Created    Finished   Due        Subject
-----------------------------------------------
1                                      test2 due:2023-06-15 spent:40 !hidedue:abc !hidedue:???? spent:40
```

As you see, now column `Due` is empty but the tag in the subject still there. And I tried to replace `hidedue`, instead I got duplicated tag with new and old values.

**Comment by aniketgm** [comment]
`$input` is what I was looking for. No wonder it was not working. thanks.
For one of the bugs you've encountered. I think the reason for the duplicated value maybe because the description already contains the tag value and then is added again.

My config below:
```
[global]
filename = "C:\\Users\\aniketgm\\MiscPrograms\\ttdl\\todo.txt"
creation_date_auto = true
fields = "prj,pri,spent"
sort = "pri"
shell = ["powershell", "-F"]
script_ext = ".ps1"
script_prefix = "C:/Users/aniketgm/MiscPrograms/ttdl/ttdlscripts/"
```

Earlier I was not able to see how the json output looked, for (`hidedue`) script. Now I could see it and I have some **_suggestion_** (not specific to hidedue but in general) [Let me know what you think about this Or if something like this is already done]:

- The description json should only contain description and no entries of, projects, context, etc. Those things can be added externally by the code when displaying on the screen (based on configuration). The output I got when playing around:
```
# P Spent  Project Subject
---------------------------
Failed to parse output of plugin ttdl-hidedue: Unexpected character: ! at (1:1)
{
    "description":  "Prepare design for Scheduler feature +myapp @ui due:2023-06-24 !hidedue:all",
    ....
}
```
As per above, the description should only contain "Prepare design for Scheduler feature", others added by code to display on screen.

For one the case (the 2 main troubles) mentioned in #75,

>  If you add a task order online on loc:amazon a xmas gift by due:2022-12-15, ttdl list shows it as order online on a xmas gift by loc:amazon. As you see, its meaning becomes a bit hard to get

For this case, Is it possible to add a place holder in the description ? so that it can be replaced by tags later. For specialTags containing exclamation "!", one knows that those are for script purposes and can be hidden (by a config explained below). I'm saying this because it impacts the text where let's say someone don't want to see tags like "!hidedue" or anything else and just the description only.

By placeholder I mean like, In python, one can add placeholders for e.g. `print("Hello {0}, Do you live in {1}", "Vladimir", "USA")`; `{0}` and `{1}`, being placeholders. Maybe Rust has something similar ? I'm not aware.

The below one is debatable. At this point I suggest to only have this as something to think about.
- Let's say a user doesn't want to see an entry at all. for e.g. I don't want to see `context` and `due` at all, no column nor in subject, then hiding it can be made configurable by adding something like `hide_entries = "due,ctx"` in the config, which will not add the those entries in description as well as hide the respective columns. Now there is one ambiguity here; what if the configs `fields` and `hide_entries` both have `due`. I leave it to you to give priority to which config, whether to hide Or show an entry in such cases. One config overrides other.

Let me know what you think about all of the above.

**Reply by VladimirMarkelov** [reply]
> For this case, Is it possible to add a place holder in the description ?

I am not sure how it should work. In case of numbers (e.g, `{0}`), TTDL cannot fill them properly as the array of tags may have arbitrary order. In case of something like `{due}`, it seems to me the same as simple tag with value (after I fix tag duplication on edit). If you are going to fill placeholders in your script then you can add any kind of placeholder to a subject that fits your criteria - TTDL won't remove any placeholder. The only trouble I see is when placeholders are used in a task that does not have any tag starting with `!`. I do not know what to do in this case - just filling placeholder with tags values seems useless as we get duplicated info in the subject.

> For one the case (the 2 main troubles) mentioned in https://github.com/VladimirMarkelov/ttdl/issues/75,

In that issue the trouble was duplicated info in column and in subject. Now you can hide tags that are shown in their own columns. So, it is a bit different issue. You suggest to hide some tags always. I agree, it may make sense. E.g, if you use context/project only for filtering when listing tasks but you do not want to see them in the output.

> Now there is one ambiguity here; what if the configs `fields` and `hide_entries` both have due

As far as I understand, `hide_entries` hides tags only inside a subject. But `fields` is the list of columns to show. So, there is no ambiguity. A user can set `fields=due` and `hide_tags=due` - in this case, it would work the same way as when passing `fields=due` and `--clean-subject=all`

**Reply by aniketgm** [reply]
@VladimirMarkelov The placeholder was a suggestion for the special use-case that was mentioned in the issue #75. Let's ignore that.

My main suggestion point is that description should only contain text and not the special tags like `+someproject` or `@somecontext`, etc. and rather that having a `fields` config for showing fields, `ttdl` can have the opposite idea i.e. to hide fields. This will remove the burden of having the `--clean-subject` logic.

How would that work ? So say for some task: "Create UI for LoginApp". I will write a very simple todo item as:
`ttdl a "Create UI for LoginApp"`
As I've not mentioned any tags/project/context here. Show just the bare minimum columns:
```
# Subject
- --------
1 Create UI for LoginApp
```
Now if project/duedate is added `tddl e 1 +projABC --set-due=today`. Then show those columns and no adding it in description.
```
# Due        Project Subject
- ---------- ------- -------
1 2023-06-27 projABC Create UI for LoginApp
```
So as and when any more info is added. Show that column. A user would then, rather than deciding to see certain columns will decide to hide certain columns. For that we can have a property like `hide_info="prj"`. The name `hide_info` just came randomly to me now. It can be different. This will keep the output clean and as mentioned above lessen the logic of `--clean-subject`.

From what I see right now, columns are shown unnecessarily even when there is no data for that. For e.g. In config file, If I set `fields="prj,due"`, even though I don't set due date for a task, I see the columns empty. Then I either have to explicitly run `ttdl --fields="<whatever field I want>"`, or change the config file.

So again what do think ? This my final note here. Don't mean to extend this any further. Discussions like this can go on to infinity. My intention is to make this better. Peaceout ✌️

**Reply by VladimirMarkelov** [reply]
Thank you for the explanation! Now I can clearly see what your idea was about.

I like your idea. Though I am not sure how to implement half of it so it looks good.

An option `hide_columns` is clear and seems useful (if we are talking about your entire idea). I do not think it makes sense to discuss about it. Let me speak about the rest.

1. The first half is easy and straightforward: automatically hide columns if no todo contains a value for the column. I agree that sometimes looking at the mostly empty output can be frustrating. I did not notice it because from the beginning I modified my config and set a minimal field list that I use almost always - so I never see an empty column in my output. With default list, it must be easier to notice.
2. The second half: automatically create columns for all the tags. It also would have been easy and straightforward if a todo list wouldn't have included some corner cases. Here the list of corner cases I am not sure how to deal with at this moment (for some of them I have an idea but it needs some refactoring and maybe a lot of work):

- Plugins. I mean tags which names start with `!`. As a workaround, TTDL can list todos twice: the first run to collect all the real tags; the second run prints out the list. It doubles the time taken by `ttdl list` command. In most cases it is not a big deal. Another workaround: do not create columns for tags starting with `!` and tags created by plugins. It is easier and takes much less time to implement. As you were the first person in years who asked "How to use plugins?" and the plugin system was broken for years, I assume nobody use plugin. In this case, I think it is safe to just ignore plugin tags and plugin output - just use a subject as-is
- We still have to double process a task list. It seems, even for a todo list without plugin calls, TTDL must process the list twice: the first run calculates the maximal width of all columns, and the following run prints everything. But processing any list in memory twice shouldn't take much time. So, the only trouble is to make changes in printing flow (refactoring)
- Not all columns are equal. Majority of values can be printed as-is. Contexts and projects needs special treatment: a todo can contain multiple context and projects. Creating a new column for each of them is not good at all. So, something must extract and print all of them in two columns: "Project" and "Context". Fortunately, this piece of code already exists and can be reused. It was better part. Now the worse part: tags. There are tags which are printed in a human-friendly way. At this moment I can remember dates(`due` and `threshold`. There are also `created` and `finished` but they are not tags, they are attributes and are processed separately) and duration(`spent`). If we want to print user-defined tags in a nice way, we have to define tag types somehow. E.g, it can be the first character of a tag value - no special character means the value is printed as-is which is a great solution in most cases. Another workaround: make option "human-friendly" affect only built-in fields and all the rest print as-is. A quick and dirty solution that may satisfy majority of people.
- Column titles. Often tag names are abbreviations. E.g, `thr` for `Threshold`. How to print nice titles for user-defined tags? A workaround: when a user adds a new tag, the user usually knows what it means and won't be surprised when this abbreviation is displayed as a column title. The only thing TTDL can do for a user is to print the tag name capitalized.

And the last thought: I think, this behavior should be tunable. Some people may have scripts that parse TTDL ouput. When TTDL output is unstable (e.g, removing a todo can also hide a column, etc), scripts can stop working. Though, it is a tiny trouble.

The discussion can result in opening a few issues, if you think the text above makes sense 😄
